Perf: implement HttpCache service to cache Accounts.read() call

jniles opened the pull request:
**Original Text**
~~This PR introduces a new factory function for debouncing function calls.  The suggested ng-debounce library did not use promises and was cumbersome to use, thus I have implemented my own based on a SO issue.  The `Accounts.read()` method has been debounced by 2000 milliseconds to allow for things like `bhAccountSelects` to benefit from the same HTTP request instead of making one each.~~


**Updated Information**
This Pull request implements a new service called `HttpCache` that should transparently wrap any HTTP requests and ensure that the same function call will return the same result.  This is useful for smaller, more frequently called routes.

_Basic Usage:_
```js
// a GET HTTP request to some url
function read(id, options) {
  return $http.get(/*...*/);
}

// to proxy, simply wrap the function in HttpCache
const proxy = HttpCache((id, options) => read(id, options));

// You can now call proxy as
proxy(1, { limit : 10 })
  .then(() => /*... */);
```

For uncached requests, the behavior is the same a raw HTTP request.  For those that have been called multiple times, the web page no longer needs to send an extra HTTP request.

If you want to skip the cache (and bust it), simply pass `true` as the third parameter.

Closes #2892.

jniles committed:
test(accounts): initial unit test for read()

A test has been added for the read() method to show that the debounce
method will actually work.

jniles committed:
feat(accounts): debounce read request

This commit adds a debounce() call to the read request.  Since it uses
promises and Angular's $timeout(), it will require flushing the
$timeouts in tests.

Closes #2892.

jniles committed:
feat(HttpCacheService): cache HTTP requests

This commit implements the HttpCacheService, a new wrapper for any $http
get request that will cache the result for a specified number of seconds
to ensure that large payloads are not called too frequently and
overload the server.

A prototype is implement on the Accounts Service.

Closes #2892.

sfount commented on client/src/js/services/HttpCacheService.js:
```diff
@@ -0,0 +1,75 @@
+ * NOTE(@jniles) - we use $interval here to avoid slowing down protractor tests.
+ * Thanks to @sfount who debugged this in Dec 2016.
+ */
+function HttpCacheService($interval) {
```
This is a super clean implementation! Really great comments and use of ES6 :ok_hand:.

sfount commented on client/src/modules/accounts/accounts.service.js:
```diff
@@ -52,8 +58,7 @@ function AccountService(Api, bhConstants) {
   function read(id, options) {
-    return Api.read.call(this, id, options)
-      .then(handleAccounts);
+    return fetcher(id, options);
```
It took me a minute to follow this logic because all of our other `read` methods follow the pattern: 
```
return fetchSomeData().then(processSomeData);
```

With this I had to follow the callback which then involved a good understanding of how `Function.prototype.call` worked. 

Could you move the `.then(handleAccounts)` out of the callback and into the `read` method? Then the callback and HttpCache are only responsible for loading accounts and the read method handles reading and pipes into processing. 

For example: 
```js
return fetcher(id, options)
  .then(handleAccounts);
```

jniles replied:
Yeah, good call.  I'll do that.

sfount reviewed (commented):
Performance kicks :soccer:, this is awesome. 
 
I haven't had time to test this but should the HttpCacheService provide an option to bust the cache if required? I may be mistaken in this logic but because angularjs service follow the [singleton pattern](https://docs.angularjs.org/guide/services), the same instance of HttpCache will be used by any module using the AccountsService. 

As we are serialising the `...args` could it possibly return the same list of accounts even if a module has updated one of them? 

For example (I know this is obscure): 
* Accounts management pages loads all accounts
* Accounts management page updates account_1 to have text 'this is new text'
* A report module loads all accounts, 10 seconds later. As I understand it account_1 still have text 'this is old test'

Commit:
feat(HttpCache): add cache bust parameter

This commit adds a cache-busting parameter to the HttpCache.  This
allows a developer to refresh a stale cache for things like updates.

sfount reviewed (approved):
This is a good clean implementation of the feature described. 

All of the tests pass so it doesn't seem to break any functionality but I can think of a few corner cases we may have to address in the future. If this cache wrapper works well going forward we could look to add to other services that could benefit from similar performance improvements.
